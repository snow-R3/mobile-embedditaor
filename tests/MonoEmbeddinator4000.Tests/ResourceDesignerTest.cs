﻿using System;
using System.CodeDom.Compiler;
using System.IO;
using IKVM.Reflection;
using Microsoft.CSharp;
using NUnit.Framework;

namespace MonoEmbeddinator4000.Tests
{
    /// <summary>
    /// These are a set of integration/approval tests validating that we are getting expected C# code from ResourceDesignerGenerator
    /// </summary>
    [TestFixture]
    public class ResourceDesignerTest
    {
        ResourceDesignerGenerator generator;
        Universe universe;

        [SetUp]
        public void SetUp()
        {
            universe = new Universe();

            generator = new ResourceDesignerGenerator
            {
                OutputDirectory = Environment.CurrentDirectory,
                PackageName = "com.mono.embeddinator",
            };
        }

        [TearDown]
        public void TearDown()
        {
            //Locks files on Windows
            universe.Dispose();

            //Temp file
            File.Delete(generator.MainAssembly);
        }

        Assembly LoadAssembly(string resourceFile)
        {
            var temp = Path.GetTempFileName();
            var parameters = new CompilerParameters
            {
                OutputAssembly = temp,
            };
            parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("System.dll"));
            parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("Mono.Android.dll"));

            AssemblyGenerator.CreateFromResource(resourceFile, parameters);

            foreach (var reference in parameters.ReferencedAssemblies)
            {
                universe.LoadFile(reference);
            }
            return universe.LoadFile(temp);
        }

        void LoadAndGenerate(string resourceFile)
        {
            var assembly = LoadAssembly(resourceFile);
            generator.Assemblies = new[] { assembly };
            generator.MainAssembly = assembly.Location;
            generator.Generate();
        }

        [Test]
        public void String()
        {
            LoadAndGenerate("String");

            string source = generator.ToSource();
            Approvals.Verify(source);
        }

        [Test]
        public void Full()
        {
            LoadAndGenerate("Full");

            string source = generator.ToSource();
            Approvals.Verify(source);
        }
    }
}
