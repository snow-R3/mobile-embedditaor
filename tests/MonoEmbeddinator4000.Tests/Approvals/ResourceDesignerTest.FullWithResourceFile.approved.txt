//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

[assembly: Android.Runtime.ResourceDesignerAttribute("__embeddinator__.Resource", IsApplication=true)]

namespace @__embeddinator__ {
    using System;
    using Android.Runtime;
    
    
    [System.CodeDom.Compiler.GeneratedCodeAttribute("Xamarin.Android.Build.Tasks", "1.0.0.0")]
    public class Resource {
        
        private static Int32 ReadFieldInt(IntPtr R, String fieldName) {
            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
            return JNIEnv.GetStaticIntField(R, fieldId);
        }
        
        private static Int32[] ReadFieldArray(IntPtr R, String fieldName) {
            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
            IntPtr value = JNIEnv.GetStaticObjectField(R, fieldId);
            return JNIEnv.GetArray<Int32>(value);
        }
        
        public static void UpdateIdValues() {
            IntPtr R;
            R = JNIEnv.FindClass("com.mono.embeddinator.R$attr");
            managedandroid.Resource.Attribute.hello = ReadFieldInt(R, "hello");
            R = JNIEnv.FindClass("com.mono.embeddinator.R$id");
            managedandroid.Resource.Id.text = ReadFieldInt(R, "text");
            R = JNIEnv.FindClass("com.mono.embeddinator.R$layout");
            managedandroid.Resource.Layout.hello = ReadFieldInt(R, "hello");
        }
    }
}
